test/e2e/azure_test.go

```diff
@@ -617,6 +617,50 @@ var _ = Describe("Workload cluster creation", func() {
 				})
 			})
 		})
+
+		It("upgrades to a new Kubernetes version", func() {
```

Comment:
Is this an entirely new cluster test (build + upgrade)?

Are we not able to add the `By("Upgrading the Kubernetes version of the cluster", func() {` somewhere in the existing AKS E2E flow? Or are we simply restoring the original cluster state by re-applying the aks flavor spec?

Reply:
Yes, I made this a separate newly-built cluster to help ensure this doesn't affect the other tests and since the CAPI upgrade tests seem to be set up similarly. AFAICT AKS doesn't allow downgrades, even between patch versions, so I don't think this is something we can reverse easily. I suppose it wouldn't be disruptive to any of the existing tests as long as this is done last.

Reply:
I do like the idea of this being a test spec we can run on any existing cluster configuration similar to what you did with the IP prefix spec, so we don't build one net new cluster for each feature we test, and we can run the same feature test across multiple cluster flavors/shapes. FWIW the self-managed cluster tests weren't structured this way but don't think they should be used as model, we should refactor them to do the same.

Reply:
That being said it doesn't prevent us from skipping the upgrade "spec" depending on the job, for example if we find it too lengthy/costly to run on every PR test

Reply:
+1

The main concern we have to have here is that we can't move the "k8s version to build a cluster with" too quickly, to ensure we always have overhead to upgrade to.

And because of that we might consider generalizing any k8s functional tests we do against the clusters so that we can get some kind of "this basic stuff works w/ 1.23 *and* 1.24" signal. Not sure if any of our current tests are sensitive to the specific Kubernetes version, so maybe just a thought going forward.

Reply:
> The main concern we have to have here is that we can't move the "k8s version to build a cluster with" too quickly, to ensure we always have overhead to upgrade to.

Since we use "latest" for the Kubernetes version, this is already an issue if we want to run this test in the same cluster. Is it acceptable to move the test down to 1.23 from 1.24 to facilitate this change?

Reply:
(IMO) yes, that's O.K., if we move the upgrade flow to happen immediatly after that "latest -1" cluster is built

Reply:
Makes sense, just pushed that change.